lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4587,10 +4577,124 @@ void HSADispatch::overrideAcquireFenceIfNeeded()
+    const uint16_t* workgroup_max_dim = device->getWorkgroupMaxDim();
+
+    int workgroup_size[3] = { 1, 1, 1};
+
+    // Check whether the user specified a workgroup size
+    if (localDims[0] != 0) {
+      for (int i = 0; i < dims; i++) {
+        // If user specify a group size that exceeds the device limit
+        // throw an error
+        if (localDims[i] > workgroup_max_dim[i]) {
+          std::stringstream msg;
+          msg << "The extent of the tile (" << localDims[i] 
+              << ") exceeds the device limit (" << workgroup_max_dim[i] << ").";
+          throw Kalmar::runtime_exception(msg.str().c_str(), -1);
+        }
+        workgroup_size[i] = localDims[i];
+      }
+    }
+    else {
+
+      constexpr unsigned int recommended_flat_workgroup_size = 64;
+
+      // user didn't specify a workgroup size
+      if (dims == 1) {
+        workgroup_size[0] = recommended_flat_workgroup_size;
+      }
+      else if (dims == 2) {
+
+        // compute the group size for the 1st dimension
+        for (unsigned int i = 1; ; i<<=1) {
+          if (i == recommended_flat_workgroup_size
+              || i >= globalDims[0]) {
+            workgroup_size[0] = i;
+            break;
+          }
+        }
+
+        // compute the group size for the 2nd dimension
+        for (unsigned int j = 1; ; j<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * j;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || j >= globalDims[1]) {
+            workgroup_size[1] = j;
+            break;
+          }
+        }
+      }
+      else if (dims == 3) {
+
+        // compute the group size for the 1st dimension
+        for (unsigned int i = 1; ; i<<=1) {
+          if (i == recommended_flat_workgroup_size
+              || i >= globalDims[0]) {
+            workgroup_size[0] = i;
+            break;
+          }
+        }
+
+        // compute the group size for the 2nd dimension
+        for (unsigned int j = 1; ; j<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * j;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || j >= globalDims[1]) {
+            workgroup_size[1] = j;
+            break;
+          }
+        }
+
+        // compute the group size for the 3rd dimension
+        for (unsigned int k = 1; ; k<<=1) {
+          unsigned int flat_group_size = workgroup_size[0] * workgroup_size[1] * k;
+          if (flat_group_size == recommended_flat_workgroup_size
+              || k >= globalDims[2]) {
+            workgroup_size[2] = k;
+            break;
+          }
+        }
+      }
+    }
+
+    auto kernel = this->kernel;
+
+    auto calculate_kernel_max_flat_workgroup_size = [&] {
+      constexpr unsigned int max_num_vgprs_per_work_item = 256;
+      constexpr unsigned int num_work_items_per_simd = 64;
+      constexpr unsigned int num_simds_per_cu = 4;
```

Comment:
Are there any constant global variables that hold these constants? I am wondering if other functions also need to use these values, they can't reference this lambda here.

Reply:
I don't see a need in this PR but it could be refactored in the future